### PR TITLE
refactor(nervous_system): Add `controller` and `hotkeys` fields to CfParticipant, CfNeuron, and CfInvestment

### PR DESCRIPTION
**Pull request by anchpop**
## Change 1: 

A `controller` field has been added to replace `hotkey_principal` (which is now deprecated and will hopefully be removed). New versions of the code populate both fields. New versions of the code only require at least one of the two fields to be populated.

This change is useful because the previous name for `controller`, `hotkey_principal`, is very confusing because it doesn't refer to the NNS neuron's hotkey, instead it refers to the NNS neuron's controller. (It is named that way because it becomes a hotkey on the SNS neuron that is allocated to that NNS neuron, but it does not represent a hotkey on the NNS neuron.)

## Change 2: 

A `hotkeys` field has been added.

This is useful because we want to transfer NF neurons' hotkeys from the NNS over to the SNS.

## Upgrade dependency

**None.** This MR merely extends various API with additional fields, and doesn’t change existing fields. All the new fields are optional, so there is no upgrade dependency introduced by this MR. No canisters currently require these fields to be `Some`.

NNS1-3198 (which tracks the removal `hotkey_principal`) can only be implemented once the old fields are no longer used by any canisters, and will have to be rolled out carefully since the field is non-optional. 

## Misc.

1. I moved the proto definition of `Principals` to nervous_system/proto. I think this is backwards compatible but the buf backwards compatibility checker disagrees. Since we haven't deployed anything that uses Principals yet, it should still be safe to merge

### CODE REVIEW

File: rs/sns/swap/tests/swap.rs [hunk 1]
```diff
@@ -3939,19 +3943,23 @@ fn test_create_sns_neuron_recipes_skips_already_created_neuron_recipes_for_nf_pa
...
             },
             CfParticipant {
-                hotkey_principal: i2principal_id_string(1002),
+                controller: Some(PrincipalId::new_user_test_id(1002)),
+                hotkey_principal: PrincipalId::new_user_test_id(1002).to_string(),
                 cf_neurons: vec![CfNeuron {
                     nns_neuron_id: 2,
                     amount_icp_e8s: 50 * E8,
                     has_created_neuron_recipes: Some(false),
+                    hotkeys: Some(Vec::new().into()),
```
Comment (aterga) [inline rs/sns/swap/tests/swap.rs:3962]: We agreed to use From instead of Into

File: rs/sns/swap/tests/swap.rs [hunk 2]
```diff
@@ -3939,19 +3943,23 @@ fn test_create_sns_neuron_recipes_skips_already_created_neuron_recipes_for_nf_pa
...
         cf_participants: vec![
             CfParticipant {
-                hotkey_principal: i2principal_id_string(1001),
+                controller: Some(PrincipalId::new_user_test_id(1001)),
+                hotkey_principal: PrincipalId::new_user_test_id(1001).to_string(),
                 cf_neurons: vec![CfNeuron {
                     nns_neuron_id: 1,
                     amount_icp_e8s: 50 * E8,
                     has_created_neuron_recipes: Some(true),
+                    hotkeys: Some(Vec::new().into()),
```
Comment (aterga) [inline rs/sns/swap/tests/swap.rs:3952]: Ditto

File: rs/sns/swap/tests/swap.rs [hunk 3]
```diff
@@ -3462,6 +3463,8 @@ async fn test_claim_swap_neuron_correctly_creates_neuron_parameters() {
                     followees: vec![NeuronId::new_test_neuron_id(20).into()],
                 }),
                 investor: Some(Investor::CommunityFund(CfInvestment {
+                    controller: Some(*TEST_USER2_PRINCIPAL),
+                    hotkeys: Some(Vec::new().into()),
```
Comment (aterga) [inline rs/sns/swap/tests/swap.rs:3467]: Ditto

File: rs/sns/swap/tests/common/mod.rs [hunk 4]
```diff
@@ -240,9 +240,22 @@ pub fn create_generic_sns_neuron_recipes(count: u64) -> Vec<SnsNeuronRecipe> {
 
 pub fn create_generic_cf_participants(count: u64) -> Vec<CfParticipant> {
     (1..count + 1)
-        .map(|i| CfParticipant {
-            hotkey_principal: i2principal_id_string(i),
-            cf_neurons: vec![CfNeuron::try_new(i, E8).unwrap()],
+        .map(|i| {
+            let i = i * 3;
```
Comment (aterga) [inline rs/sns/swap/tests/common/mod.rs:244]: Why is this needed?

File: rs/sns/swap/src/types.rs [hunk 5]
```diff
@@ -765,11 +770,16 @@ impl CfParticipant {
...
+    pub fn try_new(
+        nns_neuron_id: u64,
+        amount_icp_e8s: u64,
+        hotkeys: Vec<PrincipalId>,
+    ) -> Result<Self, String> {
         let cf_neuron = Self {
             nns_neuron_id,
             amount_icp_e8s,
             has_created_neuron_recipes: Some(false),
+            hotkeys: Some(hotkeys.into()),
```
Comment (aterga) [inline rs/sns/swap/src/types.rs:782]: From, not Into

File: rs/sns/swap/src/swap.rs [hunk 6]
```diff
@@ -3453,8 +3477,11 @@ fn create_sns_neuron_basket_for_cf_participant(
                 transfer_fee_paid_e8s: Some(0),
             }),
             investor: Some(Investor::CommunityFund(CfInvestment {
-                hotkey_principal: hotkey_principal.to_string(),
+                controller: Some(*controller),
+                hotkeys: Some(hotkeys.clone().into()),
```
Comment (aterga) [inline rs/sns/swap/src/swap.rs:3481]: Ditto

File: rs/sns/swap/src/swap.rs [hunk 7]
```diff
@@ -1641,19 +1648,25 @@ impl Swap {
...
+                            // or if neither is set, something has gone wrong and we increment the invalid field.
+                            // This will require a manual intervention via an upgrade to correct.
+                            (Some(_), Some(_)) | (None, None) => {
+                                log!(ERROR, "Invalid NF neuron: recipe={recipe:?}");
+                                sweep_result.invalid += 1;
+                                continue;
+                            }
+                            (Some(p), _) => *p,
+                            (_, Some(p)) => p,
+                        };
```
Comment (anchpop) [inline rs/sns/swap/src/swap.rs:1669]: Try clarifying this code

File: rs/sns/swap/tests/swap.rs (hunk 1 above)
Comment (max-dfinity) [reply rs/sns/swap/tests/swap.rs:3962]: Is that decision still serving us well?

Comment (anchpop) [reply rs/sns/swap/tests/swap.rs]: Not sure but I changed it anyway for clarity

File: rs/sns/swap/tests/swap.rs (hunk 3 above)
Comment (anchpop) [reply rs/sns/swap/tests/swap.rs]: Fixed

File: rs/sns/swap/src/types.rs (hunk 5 above)
Comment (anchpop) [reply rs/sns/swap/src/types.rs]: Fixed

File: rs/sns/swap/src/swap.rs (hunk 6 above)
Comment (anchpop) [reply rs/sns/swap/src/swap.rs]: Fixed

File: rs/sns/swap/src/swap.rs (hunk 7 above)
Comment (anchpop) [reply rs/sns/swap/src/swap.rs]: I ended up not being able to clarify it much with an intermediate struct, but I was able to extract the controller-getting code to a function

File: rs/sns/swap/tests/swap.rs (hunk 2 above)
Comment (anchpop) [reply rs/sns/swap/tests/swap.rs]: Fixed

File: rs/sns/swap/tests/common/mod.rs (hunk 4 above)
Comment (anchpop) [reply rs/sns/swap/tests/common/mod.rs:244]: Otherwise some neurons will have controllers that are the hotkeys of other neurons. Now that I think of it, that isn't so bad, but my first thought was just to make them all truly unrelated neurons

File: rs/sns/swap/src/types.rs [hunk 8]
```diff
@@ -708,17 +708,63 @@ impl DirectInvestment {
...
+            // If both fields are none, something has also gone wrong.
+            (None, None) => Err(
+                "Invalid NF participant: controller and hotkey_principal are both unset"
+                    .to_string(),
+            ),
+            // If only one is set, just use that one
+            (Some(p), None) => Ok(p),
+            (None, Some(p)) => Ok(p),
+        }
+    }
```
Comment (anchpop) [inline rs/sns/swap/src/types.rs:767]: Note to the reviewer: this code will get much simpler once hotkey_principal is removed, so I think we should tolerate the duplication/hairiness for now. It allows this function to be tolerant to either controller being unset or `hotkey_principal` being invalid (e.g. an empty string), which will be very helpful when removing `hotkey_principal`.

File: rs/nns/governance/canister/governance.did [hunk 9]
```diff
@@ -11,7 +11,7 @@ type Action = variant {
   SetDefaultFollowees : SetDefaultFollowees;
   RewardNodeProviders : RewardNodeProviders;
   ManageNetworkEconomics : NetworkEconomics;
-  ApproveGenesisKyc : Principals;
+  ApproveGenesisKyc : Principals_1;
```
Comment (max-dfinity) [inline rs/nns/governance/canister/governance.did:14]: Can we fix this before we merge this?  These kinds of changes are not backwards compatible for clients that use this to generate code (yet)

Comment (anchpop) [reply rs/nns/governance/canister/governance.did]: Ok, I added a commit that fixes it

File: rs/sns/swap/tests/swap.rs [hunk 10]
```diff
@@ -874,7 +875,7 @@ fn test_scenario_happy() {
                                                 *neurons_fund_participant_principal_id,
                                             ),
                                             // TODO(NNS1-3199): Populate this field if it is relevant for this test
-                                            hotkeys: Some(vec![].into()),
+                                            hotkeys: Some(Principals::from(Vec::new())),
```
Comment (daniel-wong-dfinity-org) [inline rs/sns/swap/tests/swap.rs:878]: 🙏

File: rs/sns/audit/src/lib.rs [hunk 11]
```diff
@@ -199,17 +198,16 @@ async fn validate_neurons_fund_sns_swap_participation(
         .neurons_fund_reserves
         .unwrap()
         .neurons_fund_neuron_portions;
-    let mut initial_amounts_per_controller: BTreeMap<String, _> = BTreeMap::new();
+    let mut initial_amounts_per_controller: BTreeMap<_, _> = BTreeMap::new();
```
Comment (daniel-wong-dfinity-org) [inline rs/sns/audit/src/lib.rs:201]: Just drop the type. It is now conveying zero information.

**Review by daniel-wong-dfinity-org — COMMENTED** [review]
I have one significant suggestion (not critical): normalize.

File: rs/sns/swap/proto/ic_sns_swap/pb/v1/swap.proto [hunk 12]
```diff
@@ -507,6 +507,8 @@ message CfNeuron {
   // The amount of ICP that the Neurons' Fund invests associated
   // with this neuron.
   uint64 amount_icp_e8s = 2;
+  // The principals that can vote, propose, and follow on behalf of this neuron.
+  optional ic_nervous_system.pb.v1.Principals hotkeys = 4;
```
[the remark this answers is not in the document]
Comment (anchpop) [reply rs/sns/swap/proto/ic_sns_swap/pb/v1/swap.proto:511]: Done!

File: rs/sns/swap/proto/ic_sns_swap/pb/v1/swap.proto [hunk 13]
```diff
@@ -517,10 +519,15 @@ message CfNeuron {
 
 // Represents a Neurons' Fund participant, possibly with several neurons.
 message CfParticipant {
-  // The principal that can vote on behalf of these Neurons' Fund neurons.
-  string hotkey_principal = 1;
+  // The principal that can manage this neuron.
+  optional ic_base_types.pb.v1.PrincipalId controller = 6;
```
[the remark this answers is not in the document]
Comment (anchpop) [reply rs/sns/swap/proto/ic_sns_swap/pb/v1/swap.proto]: Nice catch, thanks. Probably a copy-paste error

File: rs/sns/swap/src/swap.rs [hunk 14]
```diff
@@ -910,14 +911,20 @@ impl Swap {
                         total_participant_icp_e8s,
                     );
 
+                    #[allow(deprecated)] // TODO(NNS1-3198): Remove once hotkey_principal is removed
                     let Some(parsed_principal) =
-                        string_to_principal(&cf_participant.hotkey_principal)
+                    // TODO(NNS1-3198): Simplify once hotkey_principal is removed
```
[the remark this answers is not in the document]
Comment (anchpop) [reply rs/sns/swap/src/swap.rs:916]: My reasoning - these are technically two separate tasks:

1. Remove `#[allow(deprecated)]` (which IIRC does not emit a warning if it is unnecessary)
2. Simplifying the match (for which a comment isn't strictly necessary because once we remove hotkey_principal it will be obvious that the match is unnecessary, but if we decide to just stop supporting it rather than removing the field it may be helpful)

File: rs/sns/swap/src/swap.rs [hunk 15]
```diff
@@ -910,14 +911,20 @@ impl Swap {
                         total_participant_icp_e8s,
                     );
 
+                    #[allow(deprecated)] // TODO(NNS1-3198): Remove once hotkey_principal is removed
                     let Some(parsed_principal) =
```
[the remark this answers is not in the document]
Comment (anchpop) [reply rs/sns/swap/src/swap.rs:915]: Good point, fixed!

File: rs/sns/swap/src/swap.rs [hunk 16]
```diff
@@ -910,14 +911,20 @@ impl Swap {
...
-                        string_to_principal(&cf_participant.hotkey_principal)
+                    // TODO(NNS1-3198): Simplify once hotkey_principal is removed
+                    cf_participant
+                        .controller
+                        .or(string_to_principal(&cf_participant.hotkey_principal))
```
[the remark this answers is not in the document]
Comment (anchpop) [reply rs/sns/swap/src/swap.rs:919]: Counterpoint: by using `.or` systematically, we don't ever have to do some dataflow analysis to make sure we've covered every case. In that sense we can just be monkeys who mechanically substitute `hotkey_principal` with `controller.or(hotkey_principal)` and know that the outcome will be code that behaves correct. If this were going to live for longer, I would agree with making the code clearer by doing what you suggested, but since this will be cleaned up by the ticket linked in the comment I think the benefit is lower than the cost of making sure we don't miss a case where we should have normalized and didn't.

If we wanted to go the normalization route, I would suggest something that I think would be even better and that we should do anyway: separate our API types from the types we use for internal logic, have the API type only have a non-optional `controller` field, and do this normalization at the point of conversion. Then any code that uses the internal-logic type would be guaranteed at the type level to have an unambiguous `controller`.

File: rs/sns/swap/tests/swap.rs (hunk 10 above)
Comment (anchpop) [reply rs/sns/swap/tests/swap.rs:878]: :D

File: rs/sns/audit/src/lib.rs (hunk 11 above)
Comment (anchpop) [reply rs/sns/audit/src/lib.rs]: Thanks, good catch

File: rs/sns/swap/src/swap.rs [hunk 17]
```diff
@@ -3996,22 +4015,47 @@ mod tests {
...
+        #[allow(deprecated)] // TODO(NNS1-3198): Remove once hotkey_principal is removed
         let cf_participants = vec![
             CfParticipant {
+                controller: Some(PrincipalId::new_user_test_id(992899)),
                 hotkey_principal: PrincipalId::new_user_test_id(992899).to_string(),
-                cf_neurons: vec![CfNeuron::try_new(1, 698047).unwrap()],
+                cf_neurons: vec![CfNeuron::try_new(
+                    1,
+                    698047,
+                    Vec::new(), // TODO(NNS1-3199): Populate if relevant to this test
```
[the remark this answers is not in the document]
Comment (anchpop) [reply rs/sns/swap/src/swap.rs:4026]: I agree, it's one of the biggest oversights of rust

File: rs/sns/swap/src/swap.rs [hunk 18]
```diff
@@ -2299,6 +2302,8 @@ impl Swap {
                     }
                 }
                 Some(Investor::CommunityFund(CfInvestment {
+                    controller: _,
+                    hotkeys: _,
                     hotkey_principal: _,
                     nns_neuron_id: _,
                 })) => compute_neuron_staking_subaccount_bytes(nns_governance.into(), neuron_memo),
```
[the remark this answers is not in the document]
Comment (anchpop) [reply rs/sns/swap/src/swap.rs]: Fixed!

File: rs/sns/swap/src/swap.rs [hunk 19]
```diff
@@ -910,14 +911,20 @@ impl Swap {
...
+                        .controller
+                        .or(string_to_principal(&cf_participant.hotkey_principal))
                     else {
                         sweep_result.invalid += neuron_basket_construction_parameters.count as u32;
                         return;
                     };
                     match create_sns_neuron_basket_for_cf_participant(
                         &parsed_principal,
+                        // TODO(NNS1-3199): Populate this field
+                        Vec::new(),
```
[the remark this answers is not in the document]
Comment (anchpop) [reply rs/sns/swap/src/swap.rs:927]: As you alluded to in a previous comment, CfParticipant corresponds to a single controller principal and each CfInvestment corresponds to one neuron controlled by that principal, each having its own set of hotkeys. Or did I misunderstand you?

File: rs/sns/swap/src/swap.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (anchpop) [reply rs/sns/swap/src/swap.rs:919]: Good point, fixed

[the remark this answers is not in the document]
Comment (anchpop) [reply rs/sns/swap/src/swap.rs]: It took a bit of finangling to keep the behavior the same while appeasing the borrow checker, but I got it working, good call!

File: rs/sns/swap/proto/ic_sns_swap/pb/v1/swap.proto [hunk 20]
```diff
@@ -517,10 +519,15 @@ message CfNeuron {
 
 // Represents a Neurons' Fund participant, possibly with several neurons.
 message CfParticipant {
-  // The principal that can vote on behalf of these Neurons' Fund neurons.
-  string hotkey_principal = 1;
+  // The principal that can manage this neuron.
```
Comment (andrew-lee-work) [inline rs/sns/swap/proto/ic_sns_swap/pb/v1/swap.proto:522]: It would be clearer to say managing voting permissions. It shouldn't have full control.

File: rs/nns/governance/canister/governance.did (hunk 9 above)
Comment (anchpop) [reply rs/nns/governance/canister/governance.did]: Fixed!

**Comment by andrew-lee-work** [comment]
Overall, LGTM.

I think the original name hotkey_principal was supposed to clarify that it's not going to be a controller of the neuron. In keeping with that spirit, I think something like "voting_manager_principal" would be clearer than just "controller", since one might think from the name that it's the controller of the SNS neuron.

One minor additional risk is that the hotkeys will make neurons take up more space. But the max neurons issue is being addressed elsewhere.

File: rs/sns/swap/proto/ic_sns_swap/pb/v1/swap.proto (hunk 20 above)
Comment (anchpop) [reply rs/sns/swap/proto/ic_sns_swap/pb/v1/swap.proto]: It is very confusing in this case because `// The principal that can vote on behalf of these Neurons' Fund neurons.` in the original comment was referring to the SNS while `// The principal that can manage this neuron.` in the new comment is referring to the NNS. I've clarified the comment to prevent this confusion.

File: rs/sns/swap/src/swap.rs [hunk 21]
```diff
@@ -894,9 +894,24 @@ impl Swap {
...
+                        Err(e) => {
+                            log!(
+                                ERROR,
+                                "Error getting the controller for {cf_neuron:?} principal: {e}"
+                            );
+                            sweep_result.invalid +=
+                                neuron_basket_construction_parameters.count as u32;
+                            return;
+                        }
+                    };
```
Comment (anchpop) [inline rs/sns/swap/src/swap.rs:913]: Note to reviewer: this check must be done inside this loop because naively moving it outside of the loop would change how many times we increment sweep_result.invalid. (Unless we did something more complicated I guess). And if we're doing it inside the loop, we should do it in the closure I think, otherwise it makes the control flow complicated

File: rs/sns/swap/src/swap.rs [hunk 22]
```diff
@@ -894,9 +894,24 @@ impl Swap {
         // for all NF investors.
         let mut global_cf_memo: u64 = NEURON_BASKET_MEMO_RANGE_START;
         for cf_participant in self.cf_participants.iter_mut() {
+            let controller = cf_participant.try_get_controller();
```
Comment (anchpop) [inline rs/sns/swap/src/swap.rs:897]: Note to reviewer: This must be done outside of the loop or we get complains from borrowck

**Review by max-dfinity — APPROVED** [review]

**Review by daniel-wong-dfinity-org — APPROVED** [review]

File: rs/sns/swap/src/swap.rs [hunk 23]
```diff
@@ -894,9 +894,24 @@ impl Swap {
...
+
             for cf_neuron in cf_participant.cf_neurons.iter_mut() {
                 // Create a closure to ensure `global_cf_memo` is incremented in all cases
                 let mut process_cf_neuron = || {
+                    let controller = match controller.clone() {
+                        Ok(nns_neuron_controller_principal) => nns_neuron_controller_principal,
```
Comment (daniel-wong-dfinity-org) [inline rs/sns/swap/src/swap.rs:903]:
```
Ok(ok) => ok,
```

would also be acceptable

File: rs/sns/swap/src/swap.rs [hunk 24]
```diff
@@ -894,9 +894,24 @@ impl Swap {
         // for all NF investors.
         let mut global_cf_memo: u64 = NEURON_BASKET_MEMO_RANGE_START;
         for cf_participant in self.cf_participants.iter_mut() {
+            let controller = cf_participant.try_get_controller();
+
             for cf_neuron in cf_participant.cf_neurons.iter_mut() {
                 // Create a closure to ensure `global_cf_memo` is incremented in all cases
                 let mut process_cf_neuron = || {
+                    let controller = match controller.clone() {
```
Comment (daniel-wong-dfinity-org) [inline rs/sns/swap/src/swap.rs:902]: Isn't controller of type PrincipalId? And isn't that Copy? Shouldn't clippy say that clone is unecesary here?

Comment (anchpop) [reply rs/sns/swap/src/swap.rs:902]: the `controller` being matched upon is type `Result<PrincipalId, String>` which is not `Copy`
